packages/adblocker/src/engine/bucket/cosmetic.ts

```diff
@@ -364,21 +368,51 @@ export default class CosmeticFilterBucket {
       return true;
     });
 
+    const exceptions: Map<CosmeticFilter, CosmeticFilter> = new Map();
+
+    if (rules.length === 0) {
+      return {
+        rules: [],
+        matches: rules,
+        exceptions,
+      };
+    }
+
     // If we found at least one candidate, check if we have unhidden rules.
-    const disabledRules: Set<string> = new Set();
-    if (rules.length !== 0) {
-      this.unhideIndex.iterMatchingFilters(hostnameTokens, (rule: CosmeticFilter) => {
-        if (rule.match(hostname, domain) && !isFilterExcluded?.(rule)) {
-          disabledRules.add(rule.getSelector());
-        }
+    const disabledRules: Map<string, CosmeticFilter> = new Map();
+    this.unhideIndex.iterMatchingFilters(hostnameTokens, (rule: CosmeticFilter) => {
+      if (rule.match(hostname, domain) && !isFilterExcluded?.(rule)) {
+        disabledRules.set(rule.getSelector(), rule);
+      }
 
-        return true;
-      });
+      return true;
+    });
+
+    if (disabledRules.size === 0) {
+      return {
+        rules,
+        matches: rules,
+        exceptions,
+      };
     }
 
-    return rules.filter(
-      (rule) => disabledRules.size === 0 || disabledRules.has(rule.getSelector()) === false,
-    );
+    const modifications: CosmeticFilter[] = [];
+    let exception: CosmeticFilter | undefined;
+    for (const rule of rules) {
+      exception = disabledRules.get(rule.getSelector());
+
+      if (exception !== undefined) {
+        exceptions.set(rule, exception);
+      } else {
+        modifications.push(rule);
+      }
+    }
```

Comment:
It's a bit unclear to me here if we should return only exception rules which matched another rule or all of them? We have 3 types of rules in this scope:

1. All "positive" matches from the index
2. All "exception" matches from the index
3. Exception rules which matched with a positive rule (which is what we return)

We only return 1. and 3. but isn't expected that the event `filter-matched` would also trigger for exception rules which did not match a "positive" rule? (i.e. all the `disabledRules`)

This comment also applies for `getCosmeticsFilters` below.